### ISSUE TITLE
Performance question

### ISSUE THREAD

**Issue by booper** [VROOM-Project/vroom-docker#42]
I deployed the VROOM docker container on my own server ( 8vCPU, 16Gb RAM - provider DO ). Everything works just perfect, great soft. 
Then I tried and test it on 1900 Waypoints with 70 vehicles. It works but I get calculated optimized routes in 15 mins. 
I  run `docker stats `command  and see couples of interesting numbers:
<img width="903" alt="vroom1" src="https://user-images.githubusercontent.com/583898/127381588-554036a8-7b68-4a4e-b9a5-6b816037a883.png">

and then 

<img width="899" alt="vroom2" src="https://user-images.githubusercontent.com/583898/127381610-721e0c9f-48b0-4fb5-93c2-bf9ba1bfda7a.png">

if I understand correctly on the first pic osrm calculates matrix, the second one vroom got the matrix and start to optimize and build routes. am I right?

here are my questions:

1. When osrm works it doesn't use all available RAM ( or most of it ) just 2.3Gb from 16 available. I didn't make any limitations for docker neither for RAM nor CPU. Is it possible to osrm use more RAM and will it increase performance and decrease computing time for osrm?
2. the Second pic shows how vroom calculates routes, it uses ALL 8 CPUs but doesn't use RAM, is it ok? 
3. What is the best server configuration for high-loaded systems for VROOM. How to get optimized routes for less time with thousands of waypoints and hundreds of vehicles?
4. Is there any settings or options to configure docker containers to increase performance or power of hardware is the key?

**Comment by jcoupey** [comment]
This is not really Docker-related as the considerations around performance and tuning apply whatever the context.

> if I understand correctly on the first pic osrm calculates matrix, the second one vroom got the matrix and start to optimize and build routes. am I right?

Right.

For the questions:

1. The amount of RAM used by OSRM depend on the geographical extract you're using, and probably marginally on the size of the request. The question of OSRM performance vs resources is out of scope here, you should check the different pipelines OSRM offers.
2. Yes, the optimization phase is rather CPU intensive but there is not that much to hold on RAM: the biggest data structures used while solving are the matrices we keep, whose size depend on the overall number of locations in the problem.
3. Because the computing is CPU intensive and many operations are parallelized, a configuration with more CPU that allows to run more threads in parallel will make a difference. You may want to check https://github.com/VROOM-Project/vroom/issues/489 on this too.
4. Check out `-t` and `-x` flags from `vroom -h`, that's how you can speed-up things and/or adjust the trade-off between computing time and exploration depth.

**Comment by nilsnolde** [comment]
Well observed:) And you pretty much took the right conclusions. 

Let me chime in a little:
1. OSRM (like most routers) uses one core per request. You can install a middleware/proxy that accepts a OSRM matrix request, chunks it up in N portions (N = available threads), requests all chunks simultaneously from OSRM and stitches all response back to one response. There you can expect an almost inverse linear relationship between available threads and calculation time (depending a little on your threading strategy). Easiest done with their node bindings, I think there's a repo somewhere that implements smth like that.
2. Not too familiar with vroom's internals, but you actually want to use as many cores as you have to come to a better solution in less time (AFAIU). The `-x` exploration tells vroom how many to use: https://github.com/VROOM-Project/vroom/issues/489#issuecomment-854538767. Not sure what happens when `-x=5` and 8 cores available (guess it just takes longer since it has to calculate in sequence?). vroom should hardly need state (unlike routing engines) so RAM usage is low. Optimization is mostly CPU-bound.
3. For vroom only I guess the more (higher-clocked) threads you throw at it the better (max 32 though). For OSRM CPU plays a minor role, RAM is much more important. That's the most expensive combination, but maybe you can even split them over multiple servers.
4. If you're on linux docker uses what you have. What you're thinking of only exists for mac & win where docker is basically behind a VM. on linux you only _constrain_ docker's access to hardware, not increase it.

I'll close and guess @jcoupey will correct me if there's smth absurdly wrong

**Comment by nilsnolde** [comment]
hahah nice timing!

**Comment by booper** [comment]
thank you, guys. really helpful replies. went to gid deeper :)
